### ISSUE TITLE
[network] Allow for known trusted peers in untrusted networks

### ISSUE THREAD

**Issue by gregnazario** [diem/diem#6790]
# Overview
### Types of Networks
Currently, there are two types of networks:
* Trusted networks, with mutual authentication (e.g. Validator network, both sides are known and authenticated)
* Untrusted networks, server only authentication (e.g. Public network, only the node being connected to authenticates)

### Trusted Peers Discovery and Connections
With a Trusted network, we retrieve trusted peers via On-Chain discovery in the `ConnectivityManager`.  This type of discovery is built upon the idea of the Peer-to-Peer(P2P) network or an upstream network.  Thus, when we discover the trusted peers, we try to connect to trusted peers with outbound dials, and include keeping track of inbound dials from trusted peers.

This has led to us adding an outbound connection limit on the `Connectivity Manager` for full nodes to ensure they have only 3 upstream connections.

### Untrusted Peers Discovery and Connections
With an Untrusted network, we retrieve the first peers to connect to by seed peers, and do not retrieve on-chain discovery data or use it for mutual authentication.  Therefore, all inbound connections on this network are treated equally for better or worse.

# Problem
We currently are building some work arounds for providing prioritized or reliable access to the network for preferred nodes.  Additionally, we're working on adding connection limits to prevent having too many nodes connected.  On public untrusted networks, we currently don't treat known trusted peers differently from any other peer, which means that we could block known trusted nodes from connecting to the network.  This could block purposes such as fallback to public network when a node has fallen far enough behind that it isn't allowed in the trusted set.

The main issues are:
* How do we prevent known peers from being blocked?
* How do we allow known peers to given reliable access?
* How do we retrieve the known peers information without automatically making outbound calls?

# Approaches
tl;dr to provide better access to some prioritized peers, we should be able to take advantage of the current "trusted peers" infrastructure

**Comment by davidiw** [comment]
Support mutual and server-only auth at the same time -- we would upgrade to mutual auth if the incoming peer is a well known peer. Some code pointers:

https://github.com/libra/libra/blob/master/network/src/transport/mod.rs#L407

So in this case, we would make it such that HandshakeAuthMode should not be inferred

Then during the noise handshake, if a peer is within the trusted peer set, we use mutual auth and if it is not we use server-only, so this needs some tweaking: https://github.com/libra/libra/blob/master/network/src/noise/handshake.rs#L350

Finally a peer that is recognized as trusted must receive an appropriate designation so that is clear upstream: 
https://github.com/libra/libra/blob/master/network/src/transport/mod.rs#L282

maybe in Connection or ConnectionMetadata

note, all outbound connections, by definition, are trusted, only inbound connections that have mutual auth are trusted

**Comment by phlip9** [comment]
> Support mutual and server-only auth at the same time -- we would upgrade to mutual auth if the incoming peer is a well known peer. Some code pointers:

This definitely makes a lot of sense. I think we just need to always push down trusted peers to the Noise handshake and then tag new connections with trusted peer ids. The actual wire protocol is the same regardless of auth mode.

Though for the validator network, we still want to drop untrusted peers. The connectivity manager would probably also need a little tweaking.

**Comment by mimoo** [comment]
> With a Trusted network, we retrieve trusted peers via On-Chain discovery

This is not true at the start right? We need seed peers to bootstrap and get to the latest on-chain state

> This has led to us adding an outbound connection limit on the Connectivity Manager for full nodes to ensure they have only 3 upstream connections.

Not sure I got this, why limit to 3?

**Comment by bmwill** [comment]
> I would do this differently: you're a server in a server-only network; you receive the client's public key as part of the handshake; you notice it is part of the trusted peer set ; you indicate that to the above layer when you return from the handshake. I think the latter approach is simpler and doesn't require much changes.

@mimoo If you look at the referenced PR that's exactly how I've implemented this :)

**Comment by davidiw** [comment]
> > This has led to us adding an outbound connection limit on the Connectivity Manager for full nodes to ensure they have only 3 upstream connections.
> 
> Not sure I got this, why limit to 3?

Otherwise we connect to every peer that we have in our config / discovery

**Comment by davidiw** [comment]
> > Then during the noise handshake, if a peer is within the trusted peer set, we use mutual auth and if it is not we use server-only
> 
> So if I understand correctly you're proposing that a client that is also a trusted peer (so a validator) sends a special message before starting the noise handshake in order to use the mutuel_auth handshake instead of the server_only. I would do this differently: you're a server in a server-only network; you receive the client's public key as part of the handshake; you notice it is part of the trusted peer set ; you indicate that to the above layer when you return from the handshake. I think the latter approach is simpler and doesn't require much changes.

The proposal was not intended to be an abstraction since I was pretty confident it would require no modifications to the actual protocol but more the scaffolding, which Brandon successfully mastered.

**Comment by davidiw** [comment]
Merging this into #6948
